contracts/FeralfileArtworkV4_2.sol

```diff
@@ -0,0 +1,339 @@
+
+    string public artworkFileURI;
+
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        bool burnable_,
+        bool bridgeable_,
+        address signer_,
+        address vault_,
+        address costReceiver_,
+        string memory contractURI_,
+        uint256[] memory seriesIds_,
+        uint256[] memory seriesMaxSupplies_,
+        address erc20ContractAddress_
+    )
+        FeralfileExhibitionV4_1(
+            name_,
+            symbol_,
+            burnable_,
+            bridgeable_,
+            signer_,
+            vault_,
+            costReceiver_,
+            contractURI_,
+            seriesIds_,
+            seriesMaxSupplies_
+        )
+    {
+        _erc20Token = FeralfileToken(erc20ContractAddress_);
+    }
+    
+    /// @notice Mint new collection of Artwork
+    /// @dev the function iterates over the array of MintDataWithIndex to call the internal function _mintArtwork
+    /// @param data an array of MintDataWithIndex
+    function mintArtworksWithIndex(
+        MintDataWithIndex[] calldata data
+    ) external onlyAuthorized {
+        if (!mintable) {
+            revert MintNotEnabled();
+        }
+
+        for (uint256 i = 0; i < data.length; i++) {
+            _tokenIndexes[data[i].tokenId] = data[i].tokenIndex;
+            _mintArtwork(data[i].seriesId, data[i].tokenId, data[i].owner);
+        }
+    }
+
+    /// @notice override revert mint new collection of Artwork
+    function mintArtworks(MintData[] calldata) external override view onlyAuthorized {
+        revert FunctionNotSupported();
+    }
+
+    /// @notice Update the imageURI & parameters of an edition to a new value
+    function updateTokenInformation(uint256 tokenId, string calldata imageURI, bytes calldata parameters, uint256 coinAmount)
+        external
+        onlyAuthorized
+    {
+        if (!_exists(tokenId)) {
+            revert TokenIDNotFound();
+        }
+
+        TokenInfo memory tokenInfo = _tokenInfos[tokenId];
+        _tokenInfos[tokenId] = TokenInfo(imageURI, parameters, coinAmount);
+
+        // mint token for the owner
+        uint256 mintAmount = coinAmount - tokenInfo.coinAmount;
+        if (mintAmount > 0) {
+            address owner = ownerOf(tokenId);
+            _erc20Token.mint(owner, mintAmount * 10**18);
+        }
+    }
+
+    /// @notice pay to get artworks to a destination address. The pricing, costs and other details is included in the saleData
+    /// @param r_ - part of signature for validating parameters integrity
+    /// @param s_ - part of signature for validating parameters integrity
+    /// @param v_ - part of signature for validating parameters integrity
+    /// @param saleData_ - the sale data
+    function buyArtworks(
+        bytes32 r_,
+        bytes32 s_,
+        uint8 v_,
+        SaleData calldata saleData_
+    ) external payable override {
+        require(_selling, "FeralfileExhibitionV4: sale is not started");
```

Comment:
Why don't we just call `super.buyArtwork` instead? The coin transfer part could be done after that, it's OK to do them separately. 
I know using the same loop could reduce a bit gas fee but it would be just a little bit. Copy the code from the parent contract would make it harder to maintain the shared logic between those contracts. Please consider.

Reply:
we cannot call `super.buyArtwork` since the function was defined as external in the base contract. To do this we might need to change the access modifier in the base contract.